tools/bass/makefiles.py

```diff
@@ -0,0 +1,239 @@
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+# This is all very naive and will hurt pythonists' eyes.
+#
+
+import os
+import re
+import subprocess
+import warnings
+
+from .component import Component
+
+
+class Keywords(object):
+    def __init__(self):
+        self.variables = {
+            "BUILD_BITS":
+            ["NO_ARCH",
+             "32",
+             "64",
+             "32_and_64",
+             "64_and_32"],
+            "BUILD_STYLE":
+                ["ant",
+                 "attpackagemake",
+                 "cmake",
+                 "configure",
+                 "gem",
+                 "justmake",
+                 "makemaker",
+                 "meson",
+                 "ocaml",
+                 "setup.py",
+                 "waf"],
+            "MK_BITS":
+            ["NO_ARCH",
+             "32",
+             "64",
+             "32_and_64"],
+            "COMPONENT_NAME": [],
+            "COMPONENT_VERSION": [],
+            "COMPONENT_REVISION": [],
+            "COMPONENT_FMRI": [],
+            "COMPONENT_CLASSIFICATION": [],
+            "COMPONENT_SUMMARY": [],
+            "COMPONENT_PROJECT_URL": [],
+            "COMPONENT_SRC": ["$(COMPONENT_NAME)-$(COMPONENT_VERSION)"],
+            "COMPONENT_ARCHIVE": [],
+            "COMPONENT_ARCHIVE_URL": [],
+            "COMPONENT_ARCHIVE_HASH": [],
+            "COMPONENT_LICENSE": [],
+            "COMPONENT_LICENSE_FILE": []
+        }
+        self.targets = {
+            "build": [ "BUILD_$(MK_BITS)"],
+            "install": ["INSTALL_$(MK_BITS)"],
+            "test": ["TEST_$(MK_BITS)", "NO_TESTS"],
+            "system-test": ["SYSTEM_TEST_$(MK_BITS)", "SYSTEM_TESTS_NOT_IMPLEMENTED"]
+        }
+
+class Item(object):
+    def __init__(self, line=None, content=[]):
+        self.idx = line 
+        self.str = content 
+    def line(self):
+        return self.idx
+    def include_line(self):
+        return "include "+self.str[0]+"\n"
+    def variable_assignment(self, variable):
+        return "{:<24}".format(variable+"=")+self.str[0]
+    def target_definition(self, target):
+        ss = "{0:<24}{1}".format(target+":", self.str[0])
+        for s in self.str[1:]:
+            ss += "\n\t" + s
+        ss+= "\n"
+        return ss
+
+class Makefile(object):
+    def __init__(self, path=None, debug=False):
+        self.debug = debug
+        self.path = path
+        self.component = Component()
+        self.includes = []
+        self.variables = {}
+        self.targets = {}
+        makefile = os.path.join(path, 'Makefile')
+        with open(makefile, 'r') as f:
+            self.contents = f.readlines()
+        self.update()
+
+    def update(self):
+        # Construct list of keywords
```

Comment:
General comments on programming with high level languages:
- Procedures / methods should be short and have an appropriate name.
- Whenever you need a comment to explain some lines of code you should think about refactoring the code into a separate method with a sound name and thus don't need an explanation.
- Another general idea: don't mix high level and low level language constructs. Mixing levels of abstraction is a bad sign.

=> The update() method is too long and is doing too many low level things. This is hard to understand and hard to maintain.

Reply:
Right, I am fluent in C/C++ mainly and do not have use of high level languages :)
The update() method is a naive parsing I do not know how to do this can be simpler :)

Reply:
I updated the scripts to hide some of the details and polished a bit more.